uniffi_bindgen/src/bindings/swift/templates/Template-Bridging-Header.h

```diff
@@ -6,6 +6,22 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+// The following structs are used to implement the lowest level
+// of the FFI, and thus useful to multiple uniffied crates.
+// We ensure they are declared exactly once, with a header guard, UNIFFI_SHARED_H.
+#ifdef UNIFFI_SHARED_H
+    // We also try to prevent mixing versions of shared uniffi header structs
+    // If you add anything to the #else block, you must increment the version in UNIFFI_VERSION_x_y_z.
+    #ifndef UNIFFI_VERSION_0_8_0
+        #error Combining helper code from multiple versions of uniffi is not supported
+    #endif // ndef UNIFFI_VERSION_0_8_0
+#else
+    #define UNIFFI_SHARED_H
+    #define UNIFFI_VERSION_0_8_0
+    // ⚠️ Attention: If you change this #else block (ending in `#endif // def UNIFFI_SHARED_H`) ⚠️
+    // ⚠️ you must increment the version in all instance of UNIFFI_VERSION_ in this file to the ⚠️
+    // ⚠️ current uniffi version.                                                               ⚠️
```

Comment:
Hrm. I'm trying to mentally play out the potential ways this could go wrong versus the value it will provide us.

IIUC we don't have any tests to ensure that this version-bump gets done, so future-me could conceivably add something further down in the `#else` block without really registering this warning. The downside consequences if that does happen could be catastrophic, e.g. arbitrary memory unsafety security problems.

A different approach could have this constant change automatically on every UniFFI version bump, perhaps because we plumb in the current version number through the template, or because we configure `cargo release` to change it for us automatically. The downside consequences of that approach seem small by comparison: two uniffied components using two different versions of UniFFI that *could* have operated perfectly happily together, will refuse to do so.

In practice, I don't think we're going to actually have Swift code generated by two different versions of UniFFI that we want compiled together on purpose. We do strict version checks when building the Rust code, so the megazord won't build. We also generate the Swift bindings on demand so they *should* always use the same version of UniFFI.

So...on balance I think we should treat this as a build-misconfiguration-produces-catastrophic-security-issue prevention mechanism, and automatically use the actual UniFFI release version for this constant. Plumbing it in as a template variable seems like the easiest way to achieve that, although I haven't thought about the "how" in too much detail.

Alternately, if you feel strongly about decoupling this from the UniFFI release version, then I think we should go even further in that direction: name the constant something like `UNIFFI_SHARED_HEADER_v1` and create an entirely separate incrementing version space that we change on its own schedule, and that can't be accidentally confused for the crate release version.

Reply:
I think there are two sources of incompatibilities here:

1. two different libraries (e.g. `fxa-client` and `nimbus`) having bindings and scaffolding prepared by different versions of uniffi.
2. the same library having bindings prepared by one version, and scaffolding prepared by another version.

I think whatever we choose, the way megazording is working in Xcode, with `embedded-uniffi-bindgen`, 1. is effectively eliminated, and 2. is addressed by including the uniffi version in the `ComponentInterface` hash.

> I think we should deliberately choose not to expend effort on layers of b/w compat right now, […]
> we should also deliberately plan to revisit that question in future

I think I prefer you're final suggestion of using a different version space, until we revisit this in the future.

Reply:
> is addressed by including the uniffi version in the ComponentInterface hash.

This is a good point, we have several other layers of protection here. Thanks for talking it through!